pkg/odo/genericclioptions/context.go

```diff
@@ -35,218 +34,140 @@ type Context struct {
+		if FlagValueIfSet(parameters.cmd, ComponentFlagName) != "" {
+			if err = ctx.checkComponentExistsOrFail(); err != nil {
+				return nil, err
+			}
+		}
+
+		if parameters.routeAvailability {
+			isRouteSupported, err := ctx.Client.IsRouteSupported()
 			if err != nil {
 				return nil, err
 			}
-			context.EnvSpecificInfo.SetIsRouteSupported(isRouteSupported)
+			ctx.EnvSpecificInfo.SetIsRouteSupported(isRouteSupported)
 		}
-		context.LocalConfigProvider = context.EnvSpecificInfo
-		return context, nil
 	}
-	if parameters.IsNow {
-		context, err := NewContextCreatingAppIfNeeded(parameters.Cmd)
+
+	ctx.devfilePath = location.DevfileLocation(parameters.componentContext)
+	isDevfile := odoutil.CheckPathExists(ctx.devfilePath)
+	if parameters.devfile && isDevfile {
+		// Parse devfile and validate
+		devObj, err := devfile.ParseAndValidateFromFile(ctx.devfilePath)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse the devfile %s, with error: %s", ctx.devfilePath, err)
+		}
+		err = validate.ValidateDevfileData(devObj.Data)
 		if err != nil {
 			return nil, err
 		}
-		context.ComponentContext = parameters.ComponentContext
-		return context, nil
+		ctx.EnvSpecificInfo.SetDevfileObj(devObj)
 	}
-	context, err := NewContext(parameters.Cmd)
-	if err != nil {
-		return nil, err
-	}
-	context.ComponentContext = parameters.ComponentContext
-	return context, nil
-}
-
-//InitEnvInfoFromContext initializes envinfo from the context
-func (o *Context) InitEnvInfoFromContext() (err error) {
-	o.EnvSpecificInfo, err = envinfo.NewEnvSpecificInfo(o.ComponentContext)
-	if err != nil {
-		return err
-	}
-	return nil
-}
 
-// completeDevfilePath completes the devfile path from context
-func completeDevfilePath(componentContext, devfilePath string) string {
-	if len(devfilePath) > 0 {
-		return filepath.Join(componentContext, devfilePath)
-	} else {
-		return location.DevfileLocation(componentContext)
-	}
-}
-
-// NewContext creates a new Context struct populated with the current state based on flags specified for the provided command
-func NewContext(command *cobra.Command) (*Context, error) {
-	return newContext(command, false)
-}
-
-// NewContextCreatingAppIfNeeded creates a new Context struct populated with the current state based on flags specified for the
-// provided command, creating the application if none already exists
-func NewContextCreatingAppIfNeeded(command *cobra.Command) (*Context, error) {
-	return newContext(command, true)
+	return &Context{
+		internalCxt: ctx,
+	}, nil
 }
 
 // NewContextCompletion disables checking for a local configuration since when we use autocompletion on the command line, we
 // couldn't care less if there was a configuration. We only need to check the parameters.
 func NewContextCompletion(command *cobra.Command) *Context {
-	ctx, err := newContext(command, false)
+	ctx, err := New(CreateParameters{cmd: command})
 	if err != nil {
 		util.LogErrorAndExit(err, "")
 	}
```

Comment:
Can we fix this function to return the error as well so that we can avoid abrupt exit calls?

Reply:
This function is called during completion, and I'm not sure if we can modify its signature easily. I would prefer te change is part of another PR concerning completion.